<!doctype html>
<head>
  <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1, minimum-scale=1" />

    <title>xi_covutils.msa API documentation</title>
    <meta name="description" content="Functions to work with MSA files" />

  <link href='http://fonts.googleapis.com/css?family=Source+Sans+Pro:400,300' rel='stylesheet' type='text/css'>
  
  <style type="text/css">
  
* {
  box-sizing: border-box;
}
/*! normalize.css v1.1.1 | MIT License | git.io/normalize */

/* ==========================================================================
   HTML5 display definitions
   ========================================================================== */

/**
 * Correct `block` display not defined in IE 6/7/8/9 and Firefox 3.
 */

article,
aside,
details,
figcaption,
figure,
footer,
header,
hgroup,
main,
nav,
section,
summary {
    display: block;
}

/**
 * Correct `inline-block` display not defined in IE 6/7/8/9 and Firefox 3.
 */

audio,
canvas,
video {
    display: inline-block;
    *display: inline;
    *zoom: 1;
}

/**
 * Prevent modern browsers from displaying `audio` without controls.
 * Remove excess height in iOS 5 devices.
 */

audio:not([controls]) {
    display: none;
    height: 0;
}

/**
 * Address styling not present in IE 7/8/9, Firefox 3, and Safari 4.
 * Known issue: no IE 6 support.
 */

[hidden] {
    display: none;
}

/* ==========================================================================
   Base
   ========================================================================== */

/**
 * 1. Prevent system color scheme's background color being used in Firefox, IE,
 *    and Opera.
 * 2. Prevent system color scheme's text color being used in Firefox, IE, and
 *    Opera.
 * 3. Correct text resizing oddly in IE 6/7 when body `font-size` is set using
 *    `em` units.
 * 4. Prevent iOS text size adjust after orientation change, without disabling
 *    user zoom.
 */

html {
    background: #fff; /* 1 */
    color: #000; /* 2 */
    font-size: 100%; /* 3 */
    -webkit-text-size-adjust: 100%; /* 4 */
    -ms-text-size-adjust: 100%; /* 4 */
}

/**
 * Address `font-family` inconsistency between `textarea` and other form
 * elements.
 */

html,
button,
input,
select,
textarea {
    font-family: sans-serif;
}

/**
 * Address margins handled incorrectly in IE 6/7.
 */

body {
    margin: 0;
}

/* ==========================================================================
   Links
   ========================================================================== */

/**
 * Address `outline` inconsistency between Chrome and other browsers.
 */

a:focus {
    outline: thin dotted;
}

/**
 * Improve readability when focused and also mouse hovered in all browsers.
 */

a:active,
a:hover {
    outline: 0;
}

/* ==========================================================================
   Typography
   ========================================================================== */

/**
 * Address font sizes and margins set differently in IE 6/7.
 * Address font sizes within `section` and `article` in Firefox 4+, Safari 5,
 * and Chrome.
 */

h1 {
    font-size: 2em;
    margin: 0.67em 0;
}

h2 {
    font-size: 1.5em;
    margin: 0.83em 0;
}

h3 {
    font-size: 1.17em;
    margin: 1em 0;
}

h4 {
    font-size: 1em;
    margin: 1.33em 0;
}

h5 {
    font-size: 0.83em;
    margin: 1.67em 0;
}

h6 {
    font-size: 0.67em;
    margin: 2.33em 0;
}

/**
 * Address styling not present in IE 7/8/9, Safari 5, and Chrome.
 */

abbr[title] {
    border-bottom: 1px dotted;
}

/**
 * Address style set to `bolder` in Firefox 3+, Safari 4/5, and Chrome.
 */

b,
strong {
    font-weight: bold;
}

blockquote {
    margin: 1em 40px;
}

/**
 * Address styling not present in Safari 5 and Chrome.
 */

dfn {
    font-style: italic;
}

/**
 * Address differences between Firefox and other browsers.
 * Known issue: no IE 6/7 normalization.
 */

hr {
    -moz-box-sizing: content-box;
    box-sizing: content-box;
    height: 0;
}

/**
 * Address styling not present in IE 6/7/8/9.
 */

mark {
    background: #ff0;
    color: #000;
}

/**
 * Address margins set differently in IE 6/7.
 */

p,
pre {
    margin: 1em 0;
}

/**
 * Correct font family set oddly in IE 6, Safari 4/5, and Chrome.
 */

code,
kbd,
pre,
samp {
    font-family: monospace, serif;
    _font-family: 'courier new', monospace;
    font-size: 1em;
}

/**
 * Improve readability of pre-formatted text in all browsers.
 */

pre {
    white-space: pre;
    white-space: pre-wrap;
    word-wrap: break-word;
}

/**
 * Address CSS quotes not supported in IE 6/7.
 */

q {
    quotes: none;
}

/**
 * Address `quotes` property not supported in Safari 4.
 */

q:before,
q:after {
    content: '';
    content: none;
}

/**
 * Address inconsistent and variable font size in all browsers.
 */

small {
    font-size: 80%;
}

/**
 * Prevent `sub` and `sup` affecting `line-height` in all browsers.
 */

sub,
sup {
    font-size: 75%;
    line-height: 0;
    position: relative;
    vertical-align: baseline;
}

sup {
    top: -0.5em;
}

sub {
    bottom: -0.25em;
}

/* ==========================================================================
   Lists
   ========================================================================== */

/**
 * Address margins set differently in IE 6/7.
 */

dl,
menu,
ol,
ul {
    margin: 1em 0;
}

dd {
    margin: 0 0 0 40px;
}

/**
 * Address paddings set differently in IE 6/7.
 */

menu,
ol,
ul {
    padding: 0 0 0 40px;
}

/**
 * Correct list images handled incorrectly in IE 7.
 */

nav ul,
nav ol {
    list-style: none;
    list-style-image: none;
}

/* ==========================================================================
   Embedded content
   ========================================================================== */

/**
 * 1. Remove border when inside `a` element in IE 6/7/8/9 and Firefox 3.
 * 2. Improve image quality when scaled in IE 7.
 */

img {
    border: 0; /* 1 */
    -ms-interpolation-mode: bicubic; /* 2 */
}

/**
 * Correct overflow displayed oddly in IE 9.
 */

svg:not(:root) {
    overflow: hidden;
}

/* ==========================================================================
   Figures
   ========================================================================== */

/**
 * Address margin not present in IE 6/7/8/9, Safari 5, and Opera 11.
 */

figure {
    margin: 0;
}

/* ==========================================================================
   Forms
   ========================================================================== */

/**
 * Correct margin displayed oddly in IE 6/7.
 */

form {
    margin: 0;
}

/**
 * Define consistent border, margin, and padding.
 */

fieldset {
    border: 1px solid #c0c0c0;
    margin: 0 2px;
    padding: 0.35em 0.625em 0.75em;
}

/**
 * 1. Correct color not being inherited in IE 6/7/8/9.
 * 2. Correct text not wrapping in Firefox 3.
 * 3. Correct alignment displayed oddly in IE 6/7.
 */

legend {
    border: 0; /* 1 */
    padding: 0;
    white-space: normal; /* 2 */
    *margin-left: -7px; /* 3 */
}

/**
 * 1. Correct font size not being inherited in all browsers.
 * 2. Address margins set differently in IE 6/7, Firefox 3+, Safari 5,
 *    and Chrome.
 * 3. Improve appearance and consistency in all browsers.
 */

button,
input,
select,
textarea {
    font-size: 100%; /* 1 */
    margin: 0; /* 2 */
    vertical-align: baseline; /* 3 */
    *vertical-align: middle; /* 3 */
}

/**
 * Address Firefox 3+ setting `line-height` on `input` using `!important` in
 * the UA stylesheet.
 */

button,
input {
    line-height: normal;
}

/**
 * Address inconsistent `text-transform` inheritance for `button` and `select`.
 * All other form control elements do not inherit `text-transform` values.
 * Correct `button` style inheritance in Chrome, Safari 5+, and IE 6+.
 * Correct `select` style inheritance in Firefox 4+ and Opera.
 */

button,
select {
    text-transform: none;
}

/**
 * 1. Avoid the WebKit bug in Android 4.0.* where (2) destroys native `audio`
 *    and `video` controls.
 * 2. Correct inability to style clickable `input` types in iOS.
 * 3. Improve usability and consistency of cursor style between image-type
 *    `input` and others.
 * 4. Remove inner spacing in IE 7 without affecting normal text inputs.
 *    Known issue: inner spacing remains in IE 6.
 */

button,
html input[type="button"], /* 1 */
input[type="reset"],
input[type="submit"] {
    -webkit-appearance: button; /* 2 */
    cursor: pointer; /* 3 */
    *overflow: visible;  /* 4 */
}

/**
 * Re-set default cursor for disabled elements.
 */

button[disabled],
html input[disabled] {
    cursor: default;
}

/**
 * 1. Address box sizing set to content-box in IE 8/9.
 * 2. Remove excess padding in IE 8/9.
 * 3. Remove excess padding in IE 7.
 *    Known issue: excess padding remains in IE 6.
 */

input[type="checkbox"],
input[type="radio"] {
    box-sizing: border-box; /* 1 */
    padding: 0; /* 2 */
    *height: 13px; /* 3 */
    *width: 13px; /* 3 */
}

/**
 * 1. Address `appearance` set to `searchfield` in Safari 5 and Chrome.
 * 2. Address `box-sizing` set to `border-box` in Safari 5 and Chrome
 *    (include `-moz` to future-proof).
 */

input[type="search"] {
    -webkit-appearance: textfield; /* 1 */
    -moz-box-sizing: content-box;
    -webkit-box-sizing: content-box; /* 2 */
    box-sizing: content-box;
}

/**
 * Remove inner padding and search cancel button in Safari 5 and Chrome
 * on OS X.
 */

input[type="search"]::-webkit-search-cancel-button,
input[type="search"]::-webkit-search-decoration {
    -webkit-appearance: none;
}

/**
 * Remove inner padding and border in Firefox 3+.
 */

button::-moz-focus-inner,
input::-moz-focus-inner {
    border: 0;
    padding: 0;
}

/**
 * 1. Remove default vertical scrollbar in IE 6/7/8/9.
 * 2. Improve readability and alignment in all browsers.
 */

textarea {
    overflow: auto; /* 1 */
    vertical-align: top; /* 2 */
}

/* ==========================================================================
   Tables
   ========================================================================== */

/**
 * Remove most spacing between table cells.
 */

table {
    border-collapse: collapse;
    border-spacing: 0;
}

  </style>

  <style type="text/css">
  
  html, body {
    margin: 0;
    padding: 0;
    min-height: 100%;
  }
  body {
    background: #fff;
    font-family: "Source Sans Pro", "Helvetica Neueue", Helvetica, sans;
    font-weight: 300;
    font-size: 16px;
    line-height: 1.6em;
  }
  #content {
    width: 70%;
    max-width: 850px;
    float: left;
    padding: 30px 60px;
    border-left: 1px solid #ddd;
  }
  #sidebar {
    width: 25%;
    float: left;
    padding: 30px;
    overflow: hidden;
  }
  #nav {
    font-size: 130%;
    margin: 0 0 15px 0;
  }

  #top {
    display: block;
    position: fixed;
    bottom: 5px;
    left: 5px;
    font-size: .85em;
    text-transform: uppercase;
  }

  #footer {
    font-size: .75em;
    padding: 5px 30px;
    border-top: 1px solid #ddd;
    text-align: right;
  }
    #footer p {
      margin: 0 0 0 30px;
      display: inline-block;
    }

  h1, h2, h3, h4, h5 {
    font-weight: 300;
  }
  h1 {
    font-size: 2.5em;
    line-height: 1.1em;
    margin: 0 0 .50em 0;
  }

  h2 {
    font-size: 1.75em;
    margin: 1em 0 .50em 0;
  }

  h3 {
    margin: 25px 0 10px 0;
  }

  h4 {
    margin: 0;
    font-size: 105%;
  }

  a {
    color: #058;
    text-decoration: none;
    transition: color .3s ease-in-out;
  }

  a:hover {
    color: #e08524;
    transition: color .3s ease-in-out;
  }

  pre, code, .mono, .name {
    font-family: "Ubuntu Mono", "Cousine", "DejaVu Sans Mono", monospace;
  }

  .title .name {
    font-weight: bold;
  }
  .section-title {
    margin-top: 2em;
  }
  .ident {
    color: #900;
  }

  code {
    background: #f9f9f9;
  } 

  pre {
    background: #fefefe;
    border: 1px solid #ddd;
    box-shadow: 2px 2px 0 #f3f3f3;
    margin: 0 30px;
    padding: 15px 30px;
  }

  .codehilite {
    margin: 0 30px 10px 30px;
  }

    .codehilite pre {
      margin: 0;
    }
    .codehilite .err { background: #ff3300; color: #fff !important; } 

  table#module-list {
    font-size: 110%;
  }

    table#module-list tr td:first-child {
      padding-right: 10px;
      white-space: nowrap;
    }

    table#module-list td {
      vertical-align: top;
      padding-bottom: 8px;
    }

      table#module-list td p {
        margin: 0 0 7px 0;
      }

  .def {
    display: table;
  }

    .def p {
      display: table-cell;
      vertical-align: top;
      text-align: left;
    }

    .def p:first-child {
      white-space: nowrap;
    }

    .def p:last-child {
      width: 100%;
    }


  #index {
    list-style-type: none;
    margin: 0;
    padding: 0;
  }
    ul#index .class_name {
      /* font-size: 110%; */
      font-weight: bold;
    }
    #index ul {
      margin: 0;
    }

  .item {
    margin: 0 0 15px 0;
  }

    .item .class {
      margin: 0 0 25px 30px;
    }

      .item .class ul.class_list {
        margin: 0 0 20px 0;
      }

    .item .name {
      background: #fafafa;
      margin: 0;
      font-weight: bold;
      padding: 5px 10px;
      border-radius: 3px;
      display: inline-block;
      min-width: 40%;
    }
      .item .name:hover {
        background: #f6f6f6;
      }

    .item .empty_desc {
      margin: 0 0 5px 0;
      padding: 0;
    }

    .item .inheritance {
      margin: 3px 0 0 30px;
    }

    .item .inherited {
      color: #666;
    }

    .item .desc {
      padding: 0 8px;
      margin: 0;
    }

      .item .desc p {
        margin: 0 0 10px 0;
      }

    .source_cont {
      margin: 0;
      padding: 0;
    }

    .source_link a {
      background: #ffc300;
      font-weight: 400;
      font-size: .75em;
      text-transform: uppercase;
      color: #fff;
      text-shadow: 1px 1px 0 #f4b700;
      
      padding: 3px 8px;
      border-radius: 2px;
      transition: background .3s ease-in-out;
    }
      .source_link a:hover {
        background: #FF7200;
        text-shadow: none;
        transition: background .3s ease-in-out;
      }

    .source {
      display: none;
      max-height: 600px;
      overflow-y: scroll;
      margin-bottom: 15px;
    }

      .source .codehilite {
        margin: 0;
      }

  .desc h1, .desc h2, .desc h3 {
    font-size: 100% !important;
  }
  .clear {
    clear: both;
  }

  @media all and (max-width: 950px) {
    #sidebar {
      width: 35%;
    }
    #content {
      width: 65%;
    }
  }
  @media all and (max-width: 650px) {
    #top {
      display: none;
    }
    #sidebar {
      float: none;
      width: auto;
    }
    #content {
      float: none;
      width: auto;
      padding: 30px;
    }

    #index ul {
      padding: 0;
      margin-bottom: 15px;
    }
    #index ul li {
      display: inline-block;
      margin-right: 30px;
    }
    #footer {
      text-align: left;
    }
    #footer p {
      display: block;
      margin: inherit;
    }
  }

  /*****************************/

  </style>


  <style type="text/css">
  
/* ==========================================================================
   EXAMPLE Media Queries for Responsive Design.
   These examples override the primary ('mobile first') styles.
   Modify as content requires.
   ========================================================================== */

@media only screen and (min-width: 35em) {
    /* Style adjustments for viewports that meet the condition */
}

@media print,
       (-o-min-device-pixel-ratio: 5/4),
       (-webkit-min-device-pixel-ratio: 1.25),
       (min-resolution: 120dpi) {
    /* Style adjustments for high resolution devices */
}

/* ==========================================================================
   Print styles.
   Inlined to avoid required HTTP connection: h5bp.com/r
   ========================================================================== */

@media print {
    * {
        background: transparent !important;
        color: #000 !important; /* Black prints faster: h5bp.com/s */
        box-shadow: none !important;
        text-shadow: none !important;
    }

    a,
    a:visited {
        text-decoration: underline;
    }

    a[href]:after {
        content: " (" attr(href) ")";
    }

    abbr[title]:after {
        content: " (" attr(title) ")";
    }

    /*
     * Don't show links for images, or javascript/internal links
     */

    .ir a:after,
    a[href^="javascript:"]:after,
    a[href^="#"]:after {
        content: "";
    }

    pre,
    blockquote {
        border: 1px solid #999;
        page-break-inside: avoid;
    }

    thead {
        display: table-header-group; /* h5bp.com/t */
    }

    tr,
    img {
        page-break-inside: avoid;
    }

    img {
        max-width: 100% !important;
    }

    @page {
        margin: 0.5cm;
    }

    p,
    h2,
    h3 {
        orphans: 3;
        widows: 3;
    }

    h2,
    h3 {
        page-break-after: avoid;
    }
}

  </style>

  <script type="text/javascript">
  function toggle(id, $link) {
    $node = document.getElementById(id);
    if (!$node)
    return;
    if (!$node.style.display || $node.style.display == 'none') {
    $node.style.display = 'block';
    $link.innerHTML = 'Hide source &nequiv;';
    } else {
    $node.style.display = 'none';
    $link.innerHTML = 'Show source &equiv;';
    }
  }
  </script>
</head>
<body>
<a href="#" id="top">Top</a>

<div id="container">
    
  
  <div id="sidebar">
    <h1>Index</h1>
    <ul id="index">
    <li class="set"><h3><a href="#header-variables">Module variables</a></h3>
      
  <ul>
    <li class="mono"><a href="#xi_covutils.msa.PFAM_URL">PFAM_URL</a></li>
  </ul>

    </li>

    <li class="set"><h3><a href="#header-functions">Functions</a></h3>
      
  <ul>
    <li class="mono"><a href="#xi_covutils.msa.from_pfam">from_pfam</a></li>
    <li class="mono"><a href="#xi_covutils.msa.gapstrip">gapstrip</a></li>
    <li class="mono"><a href="#xi_covutils.msa.gapstrip_sequences">gapstrip_sequences</a></li>
    <li class="mono"><a href="#xi_covutils.msa.map_reference_to_sequence">map_reference_to_sequence</a></li>
    <li class="mono"><a href="#xi_covutils.msa.map_sequence_to_reference">map_sequence_to_reference</a></li>
    <li class="mono"><a href="#xi_covutils.msa.pop_reference">pop_reference</a></li>
    <li class="mono"><a href="#xi_covutils.msa.read_msa">read_msa</a></li>
    <li class="mono"><a href="#xi_covutils.msa.write_msa">write_msa</a></li>
  </ul>

    </li>


    </ul>
  </div>

    <article id="content">
      
  

  


  <header id="section-intro">
  <h1 class="title"><span class="name">xi_covutils.msa</span> module</h1>
  <p>Functions to work with MSA files</p>
  
  <p class="source_link"><a href="javascript:void(0);" onclick="toggle('source-xi_covutils.msa', this);">Show source &equiv;</a></p>
  <div id="source-xi_covutils.msa" class="source">
    <pre><code>"""
    Functions to work with MSA files
"""
from operator import add
from tempfile import mkdtemp
from os.path import join
from shutil import rmtree
import gzip
from Bio import SeqIO
from Bio.SeqRecord import SeqRecord
import requests
from xi_covutils.pdbmapper import map_to_ungapped
from xi_covutils.pdbmapper import align_sequence_to_sequence

PFAM_URL = 'https://pfam.xfam.org'

def map_reference_to_sequence(msa_file, sequence, start=1, end=None):
    '''Align the reference sequence or a substring from it to a another given sequence.
    Substring alignment is useful for paired MSA.

    Reference sequence is assumed to be the first sequence in the alignment.

        :param msa_file: Path to a fasta file.
        :param start: Index of the starting position of the MSA which will be mapped. Starting at 1.
        :param end: Index of the last position of the MSA which will be mapped. Starting at 1.
        :param sequence: An ungapped protein sequence to be used as destination of mapping.
    '''
    ref = str(SeqIO.parse(msa_file, "fasta").next().seq)
    end = end if end else len(ref)
    ref = ref[start-1: end].replace("-", "")
    return align_sequence_to_sequence(ref, sequence)

def _count_mismatches(aln_map, seq_src, seq_dst):
    seq_src = seq_src.upper()
    seq_dst = seq_dst.upper()
    matches_iter = (0 if (seq_src[x-1] == seq_dst[y-1]) else 1
                    for x, y in aln_map.items())
    return reduce(add, matches_iter)

def _count_gaps(aln_map, seq_src, seq_dst):
    src = (max(aln_map.keys())-min(aln_map.keys())+1)-len(aln_map)
    dst = (max(aln_map.values())-min(aln_map.values())+1)-len(aln_map)
    dangling_at_start = min(min(aln_map.keys()), min(aln_map.values()))-1
    dangling_at_end = min(len(seq_src)-max(aln_map.keys()), len(seq_dst)-max(aln_map.values()))
    return dst+src+dangling_at_start*2+dangling_at_end*2

def map_sequence_to_reference(
        msa_file, sequence, msa_format='fasta',
        mismatch_tolerance=float("inf"), gap_tolerance=float("inf")):
    """
    Creates a mapping from a custom ungapped sequence and the reference (first) sequence of and MSA.

    Returns a dict from positions of the custom sequence to the positions of the reference sequence.
    The values of the dict are dicts that contains the position number of the target sequence,
    the character of the custom source sequence and the character of the target sequences, under the
    keys: 'position', 'source' and 'target' respectively.

        :param msa_file: Path to a fasta file.
        :param sequence: An ungapped protein sequence to used as the source to the mapping.
    """
    handle = open(msa_file)
    reference = SeqIO.parse(handle, format=msa_format).next().seq
    handle.close()
    ungapped_ref = "".join([s for s in reference if not s == '-'])
    aligned = align_sequence_to_sequence(sequence, ungapped_ref)
    if (_count_mismatches(aligned, sequence, ungapped_ref) <= mismatch_tolerance and
            _count_gaps(aligned, sequence, ungapped_ref) <= gap_tolerance):
        ungapped_map = {v: k for k, v in map_to_ungapped(reference).items()}
        positions = {a: ungapped_map[aligned[a]] for a in aligned}
        sources = {a: sequence[a-1] for a in aligned}
        targets = {a: reference[positions[a]-1].upper() for a in aligned}
        return {a:{
            'position': positions[a],
            'source': sources[a],
            'target': targets[a],
        } for a in aligned}
    return {}


def _gapstrip_template(sequences, use_reference):
    if use_reference:
        template = [char == "-" for char in sequences[0]]
    else:
        templates = [[char == "-" for char in seq] for seq in sequences]
        template = [True for _ in xrange(len(templates[0]))]
        for temp in templates:
            template = [x and y for x, y in zip(temp, template)]
    return template

def gapstrip_sequences(sequences, use_reference=True):
    """
    Strips the gaps of list of sequences.

    All sequences are assumed to have the same length

    Returns a list of stripped sequences in the same order as the input sequences.

        :param use_reference=True: if True the first sequence is used as reference and
            any position containing a gap in it is removed from all sequences.
            If False, only columns that contains gaps en every sequence are removed.
    """
    template = _gapstrip_template(sequences, use_reference)
    return ["".join([c for t, c in zip(template, seq) if not t]) for seq in sequences]

def read_msa(msa_file, msa_format='fasta', as_dict=False):
    """
    Reads a complete msa_file.

    Return a list of tuples with with id and sequences or a dict from id to
    sequences.

        :param msa_file: the path of the input msa file.
        :param msa_format: the format of the msa file. Can be any of the
            supported by biopython.
        :param as_dict=False: is True returns a dict from id to sequence.
    """
    with open(msa_file) as handle:
        records = [(r.id, str(r.seq)) for r in SeqIO.parse(handle, msa_format)]
    if as_dict:
        return {seq_id: sequence for seq_id, sequence in records}
    return records

def write_msa(msa_data, msa_file):
    """
    Writes a msa to file.

    Only support fasta format at the moment.
    Input data can be:
    - a list of tuples of sequence id and sequence.
    - a dict from sequence id to sequence.

        :param msa_data: input sequence data.
        :param msa_file: path of the output file.
    """
    if isinstance(msa_data, list):
        seq_iterable = msa_data
    elif isinstance(msa_data, dict):
        seq_iterable = msa_data.items()
    else:
        raise ValueError("msa_data should be a list or dict")
    with open(msa_file, 'w') as handle:
        for seq_id, seq in seq_iterable:
            handle.write(">{}\n{}\n".format(seq_id, seq))

def pop_reference(msa_data, reference_id):
    """
    Puts a reference sequence as the first sequence of a msa.

    Returns a list of tuples of sequence id and sequences.

        :param msa_data: input msa_data
        :param reference_id: sequece if of the reference
    """
    if isinstance(msa_data, list):
        pass
    elif isinstance(msa_data, dict):
        msa_data = [(seq_id, seq) for seq_id, seq in msa_data.items()]
    else:
        raise ValueError("msa_data should be a list or dict")

    results = [(seq_id, seq) for seq_id, seq in msa_data
               if not seq_id == reference_id]
    first = [(seq_id, seq) for seq_id, seq in msa_data
             if seq_id == reference_id]
    if first:
        return first + results
    else:
        raise ValueError("Sequence: {} not in msa data".format(reference_id))

def gapstrip(msa_file, use_reference=True, msa_format='fasta'):
    """
    Strips the gaps of an MSA.

    Returns a list of SeqRecord objects from biopython.

        :param msa_file: the input msa file.
        :param use_reference=True: if True the first sequence is used as reference and
            any position containing a gap in it is removed from all sequences.
            If False, only columns that contains gaps en every sequence are removed.
        :param msa_format="fasta": any format recognized by Bio.SeqIO
    """
    with open(msa_file) as handle:
        records = [(r.id, str(r.seq)) for r in SeqIO.parse(handle, msa_format)]
    gs_result = gapstrip_sequences([s for _, s in records], use_reference)
    seq_ids = (i for i, _ in records)
    return [SeqRecord(id=i, seq=s) for i, s in zip(seq_ids, gs_result)]

def _download_pfam(pfam_acc, msa_type, tmp_dir):
    full_url = "{}/family/{}/alignment/{}/gzipped".format(PFAM_URL, pfam_acc, msa_type)
    request = requests.get(full_url, stream=True)
    gz_temp = join(tmp_dir, 'tmp.gz')
    with open(gz_temp, 'wb') as tmp_fh:
        for chunk in request.iter_content(chunk_size=1024):
            tmp_fh.write(chunk)
    request.close()
    return gz_temp

def _extract_pfam(compressed_file, outfile):
    try:
        with gzip.open(compressed_file, 'rb') as gz_fh:
            with open(outfile, 'wb') as file_handle:
                while True:
                    chunk = gz_fh.read(1024)
                    if not chunk:
                        break
                    file_handle.write(chunk)
        return True
    except IOError:
        return False

def from_pfam(pfam_acc, outfile, msa_type='full'):
    """
    Download an MSA from pfam database.

    Retrieves the requiered MSA for the accession given into a file.
    Returns True if the download was successful or False otherwise.

        :param pfam_acc:
        :outfile: the path to the output file
        :param msa_type='full': One of 'seed', 'full', 'rp15', 'rp35', 'rp55', 'rp75', 'uniprot', 'ncbi' or 'meta'.
    """
    tmp_dir = mkdtemp()
    gz_temp = _download_pfam(pfam_acc, msa_type, tmp_dir)
    status = _extract_pfam(gz_temp, outfile)
    rmtree(tmp_dir)
    return status
</code></pre>
  </div>

  </header>

  <section id="section-items">
    <h2 class="section-title" id="header-variables">Module variables</h2>
      <div class="item">
      <p id="xi_covutils.msa.PFAM_URL" class="name">var <span class="ident">PFAM_URL</span></p>
      
  
  <div class="source_cont">
</div>

      </div>

    <h2 class="section-title" id="header-functions">Functions</h2>
      
  <div class="item">
    <div class="name def" id="xi_covutils.msa.from_pfam">
    <p>def <span class="ident">from_pfam</span>(</p><p>pfam_acc, outfile, msa_type=&#39;full&#39;)</p>
    </div>
    

    
  
    <div class="desc"><p>Download an MSA from pfam database.</p>
<p>Retrieves the requiered MSA for the accession given into a file.
Returns True if the download was successful or False otherwise.</p>
<pre><code>:param pfam_acc:
:outfile: the path to the output file
:param msa_type='full': One of 'seed', 'full', 'rp15', 'rp35', 'rp55', 'rp75', 'uniprot', 'ncbi' or 'meta'.
</code></pre></div>
  <div class="source_cont">
  <p class="source_link"><a href="javascript:void(0);" onclick="toggle('source-xi_covutils.msa.from_pfam', this);">Show source &equiv;</a></p>
  <div id="source-xi_covutils.msa.from_pfam" class="source">
    <pre><code>def from_pfam(pfam_acc, outfile, msa_type='full'):
    """
    Download an MSA from pfam database.

    Retrieves the requiered MSA for the accession given into a file.
    Returns True if the download was successful or False otherwise.

        :param pfam_acc:
        :outfile: the path to the output file
        :param msa_type='full': One of 'seed', 'full', 'rp15', 'rp35', 'rp55', 'rp75', 'uniprot', 'ncbi' or 'meta'.
    """
    tmp_dir = mkdtemp()
    gz_temp = _download_pfam(pfam_acc, msa_type, tmp_dir)
    status = _extract_pfam(gz_temp, outfile)
    rmtree(tmp_dir)
    return status
</code></pre>
  </div>
</div>

  </div>
  
      
  <div class="item">
    <div class="name def" id="xi_covutils.msa.gapstrip">
    <p>def <span class="ident">gapstrip</span>(</p><p>msa_file, use_reference=True, msa_format=&#39;fasta&#39;)</p>
    </div>
    

    
  
    <div class="desc"><p>Strips the gaps of an MSA.</p>
<p>Returns a list of SeqRecord objects from biopython.</p>
<pre><code>:param msa_file: the input msa file.
:param use_reference=True: if True the first sequence is used as reference and
    any position containing a gap in it is removed from all sequences.
    If False, only columns that contains gaps en every sequence are removed.
:param msa_format="fasta": any format recognized by Bio.SeqIO
</code></pre></div>
  <div class="source_cont">
  <p class="source_link"><a href="javascript:void(0);" onclick="toggle('source-xi_covutils.msa.gapstrip', this);">Show source &equiv;</a></p>
  <div id="source-xi_covutils.msa.gapstrip" class="source">
    <pre><code>def gapstrip(msa_file, use_reference=True, msa_format='fasta'):
    """
    Strips the gaps of an MSA.

    Returns a list of SeqRecord objects from biopython.

        :param msa_file: the input msa file.
        :param use_reference=True: if True the first sequence is used as reference and
            any position containing a gap in it is removed from all sequences.
            If False, only columns that contains gaps en every sequence are removed.
        :param msa_format="fasta": any format recognized by Bio.SeqIO
    """
    with open(msa_file) as handle:
        records = [(r.id, str(r.seq)) for r in SeqIO.parse(handle, msa_format)]
    gs_result = gapstrip_sequences([s for _, s in records], use_reference)
    seq_ids = (i for i, _ in records)
    return [SeqRecord(id=i, seq=s) for i, s in zip(seq_ids, gs_result)]
</code></pre>
  </div>
</div>

  </div>
  
      
  <div class="item">
    <div class="name def" id="xi_covutils.msa.gapstrip_sequences">
    <p>def <span class="ident">gapstrip_sequences</span>(</p><p>sequences, use_reference=True)</p>
    </div>
    

    
  
    <div class="desc"><p>Strips the gaps of list of sequences.</p>
<p>All sequences are assumed to have the same length</p>
<p>Returns a list of stripped sequences in the same order as the input sequences.</p>
<pre><code>:param use_reference=True: if True the first sequence is used as reference and
    any position containing a gap in it is removed from all sequences.
    If False, only columns that contains gaps en every sequence are removed.
</code></pre></div>
  <div class="source_cont">
  <p class="source_link"><a href="javascript:void(0);" onclick="toggle('source-xi_covutils.msa.gapstrip_sequences', this);">Show source &equiv;</a></p>
  <div id="source-xi_covutils.msa.gapstrip_sequences" class="source">
    <pre><code>def gapstrip_sequences(sequences, use_reference=True):
    """
    Strips the gaps of list of sequences.

    All sequences are assumed to have the same length

    Returns a list of stripped sequences in the same order as the input sequences.

        :param use_reference=True: if True the first sequence is used as reference and
            any position containing a gap in it is removed from all sequences.
            If False, only columns that contains gaps en every sequence are removed.
    """
    template = _gapstrip_template(sequences, use_reference)
    return ["".join([c for t, c in zip(template, seq) if not t]) for seq in sequences]
</code></pre>
  </div>
</div>

  </div>
  
      
  <div class="item">
    <div class="name def" id="xi_covutils.msa.map_reference_to_sequence">
    <p>def <span class="ident">map_reference_to_sequence</span>(</p><p>msa_file, sequence, start=1, end=None)</p>
    </div>
    

    
  
    <div class="desc"><p>Align the reference sequence or a substring from it to a another given sequence.
Substring alignment is useful for paired MSA.</p>
<p>Reference sequence is assumed to be the first sequence in the alignment.</p>
<pre><code>:param msa_file: Path to a fasta file.
:param start: Index of the starting position of the MSA which will be mapped. Starting at 1.
:param end: Index of the last position of the MSA which will be mapped. Starting at 1.
:param sequence: An ungapped protein sequence to be used as destination of mapping.
</code></pre></div>
  <div class="source_cont">
  <p class="source_link"><a href="javascript:void(0);" onclick="toggle('source-xi_covutils.msa.map_reference_to_sequence', this);">Show source &equiv;</a></p>
  <div id="source-xi_covutils.msa.map_reference_to_sequence" class="source">
    <pre><code>def map_reference_to_sequence(msa_file, sequence, start=1, end=None):
    '''Align the reference sequence or a substring from it to a another given sequence.
    Substring alignment is useful for paired MSA.

    Reference sequence is assumed to be the first sequence in the alignment.

        :param msa_file: Path to a fasta file.
        :param start: Index of the starting position of the MSA which will be mapped. Starting at 1.
        :param end: Index of the last position of the MSA which will be mapped. Starting at 1.
        :param sequence: An ungapped protein sequence to be used as destination of mapping.
    '''
    ref = str(SeqIO.parse(msa_file, "fasta").next().seq)
    end = end if end else len(ref)
    ref = ref[start-1: end].replace("-", "")
    return align_sequence_to_sequence(ref, sequence)
</code></pre>
  </div>
</div>

  </div>
  
      
  <div class="item">
    <div class="name def" id="xi_covutils.msa.map_sequence_to_reference">
    <p>def <span class="ident">map_sequence_to_reference</span>(</p><p>msa_file, sequence, msa_format=&#39;fasta&#39;, mismatch_tolerance=inf, gap_tolerance=inf)</p>
    </div>
    

    
  
    <div class="desc"><p>Creates a mapping from a custom ungapped sequence and the reference (first) sequence of and MSA.</p>
<p>Returns a dict from positions of the custom sequence to the positions of the reference sequence.
The values of the dict are dicts that contains the position number of the target sequence,
the character of the custom source sequence and the character of the target sequences, under the
keys: 'position', 'source' and 'target' respectively.</p>
<pre><code>:param msa_file: Path to a fasta file.
:param sequence: An ungapped protein sequence to used as the source to the mapping.
</code></pre></div>
  <div class="source_cont">
  <p class="source_link"><a href="javascript:void(0);" onclick="toggle('source-xi_covutils.msa.map_sequence_to_reference', this);">Show source &equiv;</a></p>
  <div id="source-xi_covutils.msa.map_sequence_to_reference" class="source">
    <pre><code>def map_sequence_to_reference(
        msa_file, sequence, msa_format='fasta',
        mismatch_tolerance=float("inf"), gap_tolerance=float("inf")):
    """
    Creates a mapping from a custom ungapped sequence and the reference (first) sequence of and MSA.

    Returns a dict from positions of the custom sequence to the positions of the reference sequence.
    The values of the dict are dicts that contains the position number of the target sequence,
    the character of the custom source sequence and the character of the target sequences, under the
    keys: 'position', 'source' and 'target' respectively.

        :param msa_file: Path to a fasta file.
        :param sequence: An ungapped protein sequence to used as the source to the mapping.
    """
    handle = open(msa_file)
    reference = SeqIO.parse(handle, format=msa_format).next().seq
    handle.close()
    ungapped_ref = "".join([s for s in reference if not s == '-'])
    aligned = align_sequence_to_sequence(sequence, ungapped_ref)
    if (_count_mismatches(aligned, sequence, ungapped_ref) <= mismatch_tolerance and
            _count_gaps(aligned, sequence, ungapped_ref) <= gap_tolerance):
        ungapped_map = {v: k for k, v in map_to_ungapped(reference).items()}
        positions = {a: ungapped_map[aligned[a]] for a in aligned}
        sources = {a: sequence[a-1] for a in aligned}
        targets = {a: reference[positions[a]-1].upper() for a in aligned}
        return {a:{
            'position': positions[a],
            'source': sources[a],
            'target': targets[a],
        } for a in aligned}
    return {}
</code></pre>
  </div>
</div>

  </div>
  
      
  <div class="item">
    <div class="name def" id="xi_covutils.msa.pop_reference">
    <p>def <span class="ident">pop_reference</span>(</p><p>msa_data, reference_id)</p>
    </div>
    

    
  
    <div class="desc"><p>Puts a reference sequence as the first sequence of a msa.</p>
<p>Returns a list of tuples of sequence id and sequences.</p>
<pre><code>:param msa_data: input msa_data
:param reference_id: sequece if of the reference
</code></pre></div>
  <div class="source_cont">
  <p class="source_link"><a href="javascript:void(0);" onclick="toggle('source-xi_covutils.msa.pop_reference', this);">Show source &equiv;</a></p>
  <div id="source-xi_covutils.msa.pop_reference" class="source">
    <pre><code>def pop_reference(msa_data, reference_id):
    """
    Puts a reference sequence as the first sequence of a msa.

    Returns a list of tuples of sequence id and sequences.

        :param msa_data: input msa_data
        :param reference_id: sequece if of the reference
    """
    if isinstance(msa_data, list):
        pass
    elif isinstance(msa_data, dict):
        msa_data = [(seq_id, seq) for seq_id, seq in msa_data.items()]
    else:
        raise ValueError("msa_data should be a list or dict")

    results = [(seq_id, seq) for seq_id, seq in msa_data
               if not seq_id == reference_id]
    first = [(seq_id, seq) for seq_id, seq in msa_data
             if seq_id == reference_id]
    if first:
        return first + results
    else:
        raise ValueError("Sequence: {} not in msa data".format(reference_id))
</code></pre>
  </div>
</div>

  </div>
  
      
  <div class="item">
    <div class="name def" id="xi_covutils.msa.read_msa">
    <p>def <span class="ident">read_msa</span>(</p><p>msa_file, msa_format=&#39;fasta&#39;, as_dict=False)</p>
    </div>
    

    
  
    <div class="desc"><p>Reads a complete msa_file.</p>
<p>Return a list of tuples with with id and sequences or a dict from id to
sequences.</p>
<pre><code>:param msa_file: the path of the input msa file.
:param msa_format: the format of the msa file. Can be any of the
    supported by biopython.
:param as_dict=False: is True returns a dict from id to sequence.
</code></pre></div>
  <div class="source_cont">
  <p class="source_link"><a href="javascript:void(0);" onclick="toggle('source-xi_covutils.msa.read_msa', this);">Show source &equiv;</a></p>
  <div id="source-xi_covutils.msa.read_msa" class="source">
    <pre><code>def read_msa(msa_file, msa_format='fasta', as_dict=False):
    """
    Reads a complete msa_file.

    Return a list of tuples with with id and sequences or a dict from id to
    sequences.

        :param msa_file: the path of the input msa file.
        :param msa_format: the format of the msa file. Can be any of the
            supported by biopython.
        :param as_dict=False: is True returns a dict from id to sequence.
    """
    with open(msa_file) as handle:
        records = [(r.id, str(r.seq)) for r in SeqIO.parse(handle, msa_format)]
    if as_dict:
        return {seq_id: sequence for seq_id, sequence in records}
    return records
</code></pre>
  </div>
</div>

  </div>
  
      
  <div class="item">
    <div class="name def" id="xi_covutils.msa.write_msa">
    <p>def <span class="ident">write_msa</span>(</p><p>msa_data, msa_file)</p>
    </div>
    

    
  
    <div class="desc"><p>Writes a msa to file.</p>
<p>Only support fasta format at the moment.
Input data can be:
- a list of tuples of sequence id and sequence.
- a dict from sequence id to sequence.</p>
<pre><code>:param msa_data: input sequence data.
:param msa_file: path of the output file.
</code></pre></div>
  <div class="source_cont">
  <p class="source_link"><a href="javascript:void(0);" onclick="toggle('source-xi_covutils.msa.write_msa', this);">Show source &equiv;</a></p>
  <div id="source-xi_covutils.msa.write_msa" class="source">
    <pre><code>def write_msa(msa_data, msa_file):
    """
    Writes a msa to file.

    Only support fasta format at the moment.
    Input data can be:
    - a list of tuples of sequence id and sequence.
    - a dict from sequence id to sequence.

        :param msa_data: input sequence data.
        :param msa_file: path of the output file.
    """
    if isinstance(msa_data, list):
        seq_iterable = msa_data
    elif isinstance(msa_data, dict):
        seq_iterable = msa_data.items()
    else:
        raise ValueError("msa_data should be a list or dict")
    with open(msa_file, 'w') as handle:
        for seq_id, seq in seq_iterable:
            handle.write(">{}\n{}\n".format(seq_id, seq))
</code></pre>
  </div>
</div>

  </div>
  


  </section>

    </article>
  <div class="clear"> </div>
  <footer id="footer">
    <p>
      Documentation generated by
      <a href="https://github.com/BurntSushi/pdoc">pdoc 0.3.2</a>
    </p>

    <p>pdoc is in the public domain with the
      <a href="http://unlicense.org">UNLICENSE</a></p>

    <p>Design by <a href="http://nadh.in">Kailash Nadh</a></p>
  </footer>
</div>
</body>
</html>